toolkit/offchain/src/d_param/mod.rs

```diff
@@ -140,17 +147,41 @@ async fn insert_d_param<C: QueryLedgerState + Transactions + QueryNetwork>(
 	)
 	.await?;
 
-	let signed_tx = ctx.sign(&tx).to_bytes();
-	let res = client.submit_transaction(&signed_tx).await.map_err(|e| {
-		anyhow!(
-			"Submit insert D-parameter transaction request failed: {}, bytes: {}",
-			e,
-			hex::encode(signed_tx)
-		)
-	})?;
-	let tx_id = McTxHash(res.transaction.id);
-	log::info!("Transaction submitted: {}", hex::encode(tx_id.0));
-	Ok(tx_id)
+	if gov_data.policy.is_single_key_policy_for(&ctx.payment_key_hash()) {
+		let signed_tx = ctx.sign(&tx).to_bytes();
+		let res = client.submit_transaction(&signed_tx).await.map_err(|e| {
+			anyhow!(
+				"Submit insert D-parameter transaction request failed: {}, bytes: {}",
+				e,
+				hex::encode(signed_tx)
+			)
+		})?;
+		let tx_id = McTxHash(res.transaction.id);
+		log::info!("Transaction submitted: {}", hex::encode(tx_id.0));
+		Ok(TxHash(tx_id))
+	} else {
+		let authorities = gov_data.policy.key_hashes();
+		if authorities.contains(&ctx.payment_key_hash().to_bytes().try_into().unwrap()) {
```

Comment:
I would say it doesn't matter if `ctx.payment_key_hash` is an authority or not. Since it pays for the TX, it needs to sign it. I propose simplifying it by dropping this check and else case and dropping the `key_hashes()` combinator.